packages/sanity/src/core/validation/validateDocumentWithReferences.ts

```diff
@@ -0,0 +1,171 @@
+import {type SanityClient} from '@sanity/client'
+import {
+  isReference,
+  type SanityDocument,
+  type Schema,
+  type ValidationContext,
+  type ValidationMarker,
+} from '@sanity/types'
+import {reduce as reduceJSON} from 'json-reduce'
+import {
+  asyncScheduler,
+  combineLatest,
+  concat,
+  defer,
+  from,
+  lastValueFrom,
+  type Observable,
+  of,
+  timer,
+} from 'rxjs'
+import {
+  distinct,
+  distinctUntilChanged,
+  first,
+  groupBy,
+  map,
+  mergeMap,
+  scan,
+  shareReplay,
+  skip,
+  throttleTime,
+} from 'rxjs/operators'
+import {exhaustMapWithTrailing} from 'rxjs-exhaustmap-with-trailing'
+import shallowEquals from 'shallow-equals'
+
+import {type DocumentPreviewStore, type LocaleSource, type SourceClientOptions} from '..'
+import {validateDocumentObservable} from './validateDocument'
+
+/**
+ * @hidden
+ * @beta */
+export interface ValidationStatus {
+  isValidating: boolean
+  validation: ValidationMarker[]
+  revision?: string
+}
+
+const INITIAL_VALIDATION_STATUS: ValidationStatus = {
+  isValidating: true,
+  validation: [],
+}
+
+function findReferenceIds(obj: any): Set<string> {
+  return reduceJSON(
+    obj,
+    (acc, node) => {
+      if (isReference(node)) {
+        acc.add(node._ref)
+      }
+      return acc
+    },
+    new Set<string>(),
+  )
+}
+
+const EMPTY_VALIDATION: ValidationMarker[] = []
+
+type GetDocumentExists = NonNullable<ValidationContext['getDocumentExists']>
+
+const listenDocumentExists = (
+  observeDocumentAvailability: DocumentPreviewStore['unstable_observeDocumentPairAvailability'],
+  id: string,
+): Observable<boolean> =>
+  observeDocumentAvailability(id).pipe(map(({published}) => published.available))
+
+// throttle delay for referenced document updates (i.e. time between responding to changes in referenced documents)
+const REF_UPDATE_DELAY = 1000
+
+function shareLatestWithRefCount<T>() {
+  return shareReplay<T>({bufferSize: 1, refCount: true})
+}
+
+/**
+ * @internal
+ * Takes an observable of a document and validates it, including any references in the document.
+ * */
+export function validateDocumentWithReferences(
```

Comment:
Could we include the change to https://github.com/sanity-io/sanity/blob/corel/packages/sanity/src/core/store/_legacy/document/document-pair/validation.ts#L52-L53 in this pr?
This file is generated from there. It's extracting the validation logic from that file and making it reusable.

Reply:
We would need to move the copy the whole file from `corel` to here

Reply:
Yeh that one slipped through, I will add to this PR